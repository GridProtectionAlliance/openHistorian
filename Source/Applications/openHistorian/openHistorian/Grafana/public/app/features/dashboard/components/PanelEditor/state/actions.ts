import { DashboardModel, PanelModel } from '../../../state';
import { ThunkResult } from 'app/types';
import {
  closeCompleted,
  PANEL_EDITOR_UI_STATE_STORAGE_KEY,
  PanelEditorUIState,
  setPanelEditorUIState,
  updateEditorInitState,
} from './reducers';
import { cleanUpEditPanel, panelModelAndPluginReady } from '../../../state/reducers';
import store from '../../../../../core/store';

export function initPanelEditor(sourcePanel: PanelModel, dashboard: DashboardModel): ThunkResult<void> {
  return dispatch => {
    const panel = dashboard.initEditPanel(sourcePanel);

    dispatch(
      updateEditorInitState({
        panel,
        sourcePanel,
      })
    );
  };
}

export function panelEditorCleanUp(): ThunkResult<void> {
  return (dispatch, getStore) => {
    const dashboard = getStore().dashboard.getModel();
    const { getPanel, getSourcePanel, shouldDiscardChanges } = getStore().panelEditor;

    if (!shouldDiscardChanges) {
      const panel = getPanel();
      const modifiedSaveModel = panel.getSaveModel();
      const sourcePanel = getSourcePanel();
      const panelTypeChanged = sourcePanel.type !== panel.type;

      // restore the source panel id before we update source panel
      modifiedSaveModel.id = sourcePanel.id;

      sourcePanel.restoreModel(modifiedSaveModel);

      // Loaded plugin is not included in the persisted properties
      // So is not handled by restoreModel
      sourcePanel.plugin = panel.plugin;

      if (panelTypeChanged) {
        dispatch(panelModelAndPluginReady({ panelId: sourcePanel.id, plugin: panel.plugin! }));
      }

      // Resend last query result on source panel query runner
      // But do this after the panel edit editor exit process has completed
      setTimeout(() => {
        sourcePanel.getQueryRunner().useLastResultFrom(panel.getQueryRunner());
      }, 20);
    }

    if (dashboard) {
      dashboard.exitPanelEditor();
    }

    dispatch(cleanUpEditPanel());
    dispatch(closeCompleted());
  };
}

export function updatePanelEditorUIState(uiState: Partial<PanelEditorUIState>): ThunkResult<void> {
  return (dispatch, getStore) => {
    const nextState = { ...getStore().panelEditor.ui, ...uiState };
    dispatch(setPanelEditorUIState(nextState));
    try {
      store.setObject(PANEL_EDITOR_UI_STATE_STORAGE_KEY, nextState);
    } catch (error) {
      console.error(error);
    }
  };
}
