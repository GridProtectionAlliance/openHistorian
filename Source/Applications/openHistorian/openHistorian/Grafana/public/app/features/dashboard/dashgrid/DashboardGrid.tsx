// Libaries
import React, { PureComponent } from 'react';
import { hot } from 'react-hot-loader';
import ReactGridLayout, { ItemCallback } from 'react-grid-layout';
import classNames from 'classnames';
// @ts-ignore
import sizeMe from 'react-sizeme';

// Components
import { AddPanelWidget } from '../components/AddPanelWidget';
import { DashboardRow } from '../components/DashboardRow';

// Types
import { GRID_CELL_HEIGHT, GRID_CELL_VMARGIN, GRID_COLUMN_COUNT } from 'app/core/constants';
import { DashboardPanel } from './DashboardPanel';
import { DashboardModel, PanelModel } from '../state';
import { CoreEvents } from 'app/types';
import { panelAdded, panelRemoved } from '../state/PanelModel';

let lastGridWidth = 1200;
let ignoreNextWidthChange = false;

interface GridWrapperProps {
  size: { width: number };
  layout: ReactGridLayout.Layout[];
  onLayoutChange: (layout: ReactGridLayout.Layout[]) => void;
  children: JSX.Element | JSX.Element[];
  onDragStop: ItemCallback;
  onResize: ItemCallback;
  onResizeStop: ItemCallback;
  onWidthChange: () => void;
  className: string;
  isResizable?: boolean;
  isDraggable?: boolean;
  viewPanel: PanelModel | null;
}

function GridWrapper({
  size,
  layout,
  onLayoutChange,
  children,
  onDragStop,
  onResize,
  onResizeStop,
  onWidthChange,
  className,
  isResizable,
  isDraggable,
  viewPanel,
}: GridWrapperProps) {
  const width = size.width > 0 ? size.width : lastGridWidth;

  // logic to ignore width changes (optimization)
  if (width !== lastGridWidth) {
    if (ignoreNextWidthChange) {
      ignoreNextWidthChange = false;
    } else if (!viewPanel && Math.abs(width - lastGridWidth) > 8) {
      onWidthChange();
      lastGridWidth = width;
    }
  }

  /*
    Disable draggable if mobile device, solving an issue with unintentionally
     moving panels. https://github.com/grafana/grafana/issues/18497
     theme.breakpoints.md = 769
  */
  const draggable = width <= 769 ? false : isDraggable;

  return (
    <ReactGridLayout
      width={lastGridWidth}
      className={className}
      isDraggable={draggable}
      isResizable={isResizable}
      containerPadding={[0, 0]}
      useCSSTransforms={false}
      margin={[GRID_CELL_VMARGIN, GRID_CELL_VMARGIN]}
      cols={GRID_COLUMN_COUNT}
      rowHeight={GRID_CELL_HEIGHT}
      draggableHandle=".grid-drag-handle"
      layout={layout}
      onResize={onResize}
      onResizeStop={onResizeStop}
      onDragStop={onDragStop}
      onLayoutChange={onLayoutChange}
    >
      {children}
    </ReactGridLayout>
  );
}

const SizedReactLayoutGrid = sizeMe({ monitorWidth: true })(GridWrapper);

export interface Props {
  dashboard: DashboardModel;
  editPanel: PanelModel | null;
  viewPanel: PanelModel | null;
  scrollTop: number;
  isPanelEditorOpen?: boolean;
}

export class DashboardGrid extends PureComponent<Props> {
  panelMap: { [id: string]: PanelModel };
  panelRef: { [id: string]: HTMLElement } = {};

  componentDidMount() {
    const { dashboard } = this.props;

    dashboard.on(panelAdded, this.triggerForceUpdate);
    dashboard.on(panelRemoved, this.triggerForceUpdate);
    dashboard.on(CoreEvents.repeatsProcessed, this.triggerForceUpdate);
    dashboard.on(CoreEvents.rowCollapsed, this.triggerForceUpdate);
    dashboard.on(CoreEvents.rowExpanded, this.triggerForceUpdate);
  }

  componentWillUnmount() {
    const { dashboard } = this.props;
    dashboard.off(panelAdded, this.triggerForceUpdate);
    dashboard.off(panelRemoved, this.triggerForceUpdate);
    dashboard.off(CoreEvents.repeatsProcessed, this.triggerForceUpdate);
    dashboard.off(CoreEvents.rowCollapsed, this.triggerForceUpdate);
    dashboard.off(CoreEvents.rowExpanded, this.triggerForceUpdate);
  }

  buildLayout() {
    const layout = [];
    this.panelMap = {};

    for (const panel of this.props.dashboard.panels) {
      const stringId = panel.id.toString();
      this.panelMap[stringId] = panel;

      if (!panel.gridPos) {
        console.log('panel without gridpos');
        continue;
      }

      const panelPos: any = {
        i: stringId,
        x: panel.gridPos.x,
        y: panel.gridPos.y,
        w: panel.gridPos.w,
        h: panel.gridPos.h,
      };

      if (panel.type === 'row') {
        panelPos.w = GRID_COLUMN_COUNT;
        panelPos.h = 1;
        panelPos.isResizable = false;
        panelPos.isDraggable = panel.collapsed;
      }

      layout.push(panelPos);
    }

    return layout;
  }

  onLayoutChange = (newLayout: ReactGridLayout.Layout[]) => {
    for (const newPos of newLayout) {
      this.panelMap[newPos.i!].updateGridPos(newPos);
    }

    this.props.dashboard.sortPanelsByGridPos();

    // Call render() after any changes.  This is called when the layour loads
    this.forceUpdate();
  };

  triggerForceUpdate = () => {
    this.forceUpdate();
  };

  onWidthChange = () => {
    for (const panel of this.props.dashboard.panels) {
      panel.resizeDone();
    }
  };

  updateGridPos = (item: ReactGridLayout.Layout, layout: ReactGridLayout.Layout[]) => {
    this.panelMap[item.i!].updateGridPos(item);

    // react-grid-layout has a bug (#670), and onLayoutChange() is only called when the component is mounted.
    // So it's required to call it explicitly when panel resized or moved to save layout changes.
    this.onLayoutChange(layout);
  };

  onResize: ItemCallback = (layout, oldItem, newItem) => {
    this.panelMap[newItem.i!].updateGridPos(newItem);
  };

  onResizeStop: ItemCallback = (layout, oldItem, newItem) => {
    this.updateGridPos(newItem, layout);
    this.panelMap[newItem.i!].resizeDone();
  };

  onDragStop: ItemCallback = (layout, oldItem, newItem) => {
    this.updateGridPos(newItem, layout);
  };

  isInView = (panel: PanelModel): boolean => {
    if (panel.isViewing || panel.isEditing) {
      return true;
    }

    // elem is set *after* the first render
    const elem = this.panelRef[panel.id.toString()];
    if (!elem) {
      // NOTE the gridPos is also not valid until after the first render
      // since it is passed to the layout engine and made to be valid
      // for example, you can have Y=0 for everything and it will stack them
      // down vertically in the second call
      return false;
    }

    const top = elem.offsetTop;
    const height = panel.gridPos.h * GRID_CELL_HEIGHT + 40;
    const bottom = top + height;

    // Show things that are almost in the view
    const buffer = 250;

    const viewTop = this.props.scrollTop;
    if (viewTop > bottom + buffer) {
      return false; // The panel is above the viewport
    }

    // Use the whole browser height (larger than real value)
    // TODO? is there a better way
    const viewHeight = isNaN(window.innerHeight) ? (window as any).clientHeight : window.innerHeight;
    const viewBot = viewTop + viewHeight;
    if (top > viewBot + buffer) {
      return false;
    }

    return !this.props.dashboard.otherPanelInFullscreen(panel);
  };

  renderPanels() {
    const panelElements = [];

    for (const panel of this.props.dashboard.panels) {
      const panelClasses = classNames({ 'react-grid-item--fullscreen': panel.isViewing });
      const id = panel.id.toString();
      panel.isInView = this.isInView(panel);

      panelElements.push(
        <div key={id} className={panelClasses} id={'panel-' + id} ref={elem => elem && (this.panelRef[id] = elem)}>
          {this.renderPanel(panel)}
        </div>
      );
    }

    return panelElements;
  }

  renderPanel(panel: PanelModel) {
    if (panel.type === 'row') {
      return <DashboardRow panel={panel} dashboard={this.props.dashboard} />;
    }

    if (panel.type === 'add-panel') {
      return <AddPanelWidget panel={panel} dashboard={this.props.dashboard} />;
    }

    return (
      <DashboardPanel
        panel={panel}
        dashboard={this.props.dashboard}
        isEditing={panel.isEditing}
        isViewing={panel.isViewing}
        isInView={panel.isInView}
      />
    );
  }

  render() {
    const { dashboard, viewPanel } = this.props;

    return (
      <SizedReactLayoutGrid
        className={classNames({ layout: true })}
        layout={this.buildLayout()}
        isResizable={dashboard.meta.canEdit}
        isDraggable={dashboard.meta.canEdit}
        onLayoutChange={this.onLayoutChange}
        onWidthChange={this.onWidthChange}
        onDragStop={this.onDragStop}
        onResize={this.onResize}
        onResizeStop={this.onResizeStop}
        viewPanel={viewPanel}
      >
        {this.renderPanels()}
      </SizedReactLayoutGrid>
    );
  }
}

export default hot(module)(DashboardGrid);
