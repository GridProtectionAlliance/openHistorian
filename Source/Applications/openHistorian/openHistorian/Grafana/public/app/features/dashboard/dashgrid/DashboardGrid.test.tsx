import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { DashboardGrid, Props } from './DashboardGrid';
import { DashboardModel } from '../state';

interface ScenarioContext {
  props: Props;
  wrapper?: ShallowWrapper<Props, any, DashboardGrid>;
  setup?: (fn: () => void) => void;
  setProps: (props: Partial<Props>) => void;
}

function getTestDashboard(overrides?: any, metaOverrides?: any): DashboardModel {
  const data = Object.assign(
    {
      title: 'My dashboard',
      panels: [
        {
          id: 1,
          type: 'graph',
          title: 'My graph',
          gridPos: { x: 0, y: 0, w: 24, h: 10 },
        },
        {
          id: 2,
          type: 'graph2',
          title: 'My graph2',
          gridPos: { x: 0, y: 10, w: 25, h: 10 },
        },
        {
          id: 3,
          type: 'graph3',
          title: 'My graph3',
          gridPos: { x: 0, y: 20, w: 25, h: 100 },
        },
        {
          id: 4,
          type: 'graph4',
          title: 'My graph4',
          gridPos: { x: 0, y: 120, w: 25, h: 10 },
        },
      ],
    },
    overrides
  );

  const meta = Object.assign({ canSave: true, canEdit: true }, metaOverrides);
  return new DashboardModel(data, meta);
}

function dashboardGridScenario(description: string, scenarioFn: (ctx: ScenarioContext) => void) {
  describe(description, () => {
    let setupFn: () => void;

    const ctx: ScenarioContext = {
      setup: fn => {
        setupFn = fn;
      },
      props: {
        isEditing: false,
        isFullscreen: false,
        scrollTop: null,
        dashboard: getTestDashboard(),
      },
      setProps: (props: Partial<Props>) => {
        Object.assign(ctx.props, props);
        if (ctx.wrapper) {
          ctx.wrapper.setProps(ctx.props);
        }
      },
    };

    beforeEach(() => {
      setupFn();
      ctx.wrapper = shallow(<DashboardGrid {...ctx.props} />);
    });

    scenarioFn(ctx);
  });
}

describe('DashboardGrid', () => {
  dashboardGridScenario('Can render dashboard grid', ctx => {
    ctx.setup(() => {});

    it('Should render', () => {
      expect(ctx.wrapper).toMatchSnapshot();
    });
  });
});
