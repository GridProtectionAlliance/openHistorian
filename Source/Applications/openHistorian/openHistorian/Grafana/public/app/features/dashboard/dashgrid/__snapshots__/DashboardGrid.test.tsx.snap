// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardGrid Can render dashboard grid Should render 1`] = `
<SizeMe(GridWrapper)
  className="layout"
  isDraggable={true}
  isFullscreen={false}
  isResizable={true}
  layout={
    Array [
      Object {
        "h": 10,
        "i": "1",
        "w": 24,
        "x": 0,
        "y": 0,
      },
      Object {
        "h": 10,
        "i": "2",
        "w": 25,
        "x": 0,
        "y": 10,
      },
      Object {
        "h": 100,
        "i": "3",
        "w": 25,
        "x": 0,
        "y": 20,
      },
      Object {
        "h": 10,
        "i": "4",
        "w": 25,
        "x": 0,
        "y": 120,
      },
    ]
  }
  onDragStop={[Function]}
  onLayoutChange={[Function]}
  onResize={[Function]}
  onResizeStop={[Function]}
  onWidthChange={[Function]}
>
  <div
    className=""
    id="panel-1"
    key="1"
  >
    <DashboardPanel
      dashboard={
        DashboardModel {
          "annotations": Object {
            "list": Array [
              Object {
                "builtIn": 1,
                "datasource": "-- Grafana --",
                "enable": true,
                "hide": true,
                "iconColor": "rgba(0, 211, 255, 1)",
                "name": "Annotations & Alerts",
                "type": "dashboard",
              },
            ],
          },
          "autoUpdate": undefined,
          "description": undefined,
          "editable": true,
          "events": Emitter {
            "emitter": EventEmitter {
              "_events": Object {
                "panel-added": EE {
                  "context": [Circular],
                  "fn": [Function],
                  "once": false,
                },
                "panel-removed": EE {
                  "context": [Circular],
                  "fn": [Function],
                  "once": false,
                },
                "repeats-processed": EE {
                  "context": [Circular],
                  "fn": [Function],
                  "once": false,
                },
                "row-collapsed": EE {
                  "context": [Circular],
                  "fn": [Function],
                  "once": false,
                },
                "row-expanded": EE {
                  "context": [Circular],
                  "fn": [Function],
                  "once": false,
                },
                "view-mode-changed": EE {
                  "context": [Circular],
                  "fn": [Function],
                  "once": false,
                },
              },
              "_eventsCount": 6,
            },
          },
          "gnetId": null,
          "graphTooltip": 0,
          "id": null,
          "links": Array [],
          "meta": Object {
            "canEdit": true,
            "canMakeEditable": false,
            "canSave": true,
            "canShare": true,
            "canStar": true,
            "fullscreen": false,
            "isEditing": false,
            "showSettings": true,
          },
          "originalTemplating": Array [],
          "originalTime": Object {
            "from": "now-6h",
            "to": "now",
          },
          "panels": Array [
            PanelModel {
              "cachedPluginOptions": Object {},
              "datasource": null,
              "events": Emitter {
                "emitter": EventEmitter {
                  "_events": Object {},
                  "_eventsCount": 0,
                },
              },
              "gridPos": Object {
                "h": 10,
                "w": 24,
                "x": 0,
                "y": 0,
              },
              "id": 1,
              "isInView": false,
              "targets": Array [
                Object {
                  "refId": "A",
                },
              ],
              "title": "My graph",
              "transparent": false,
              "type": "graph",
            },
            PanelModel {
              "cachedPluginOptions": Object {},
              "datasource": null,
              "events": Emitter {
                "emitter": EventEmitter {
                  "_events": Object {},
                  "_eventsCount": 0,
                },
              },
              "gridPos": Object {
                "h": 10,
                "w": 25,
                "x": 0,
                "y": 10,
              },
              "id": 2,
              "isInView": false,
              "targets": Array [
                Object {
                  "refId": "A",
                },
              ],
              "title": "My graph2",
              "transparent": false,
              "type": "graph2",
            },
            PanelModel {
              "cachedPluginOptions": Object {},
              "datasource": null,
              "events": Emitter {
                "emitter": EventEmitter {
                  "_events": Object {},
                  "_eventsCount": 0,
                },
              },
              "gridPos": Object {
                "h": 100,
                "w": 25,
                "x": 0,
                "y": 20,
              },
              "id": 3,
              "isInView": false,
              "targets": Array [
                Object {
                  "refId": "A",
                },
              ],
              "title": "My graph3",
              "transparent": false,
              "type": "graph3",
            },
            PanelModel {
              "cachedPluginOptions": Object {},
              "datasource": null,
              "events": Emitter {
                "emitter": EventEmitter {
                  "_events": Object {},
                  "_eventsCount": 0,
                },
              },
              "gridPos": Object {
                "h": 10,
                "w": 25,
                "x": 0,
                "y": 120,
              },
              "id": 4,
              "isInView": false,
              "targets": Array [
                Object {
                  "refId": "A",
                },
              ],
              "title": "My graph4",
              "transparent": false,
              "type": "graph4",
            },
          ],
          "refresh": undefined,
          "revision": undefined,
          "schemaVersion": 22,
          "snapshot": undefined,
          "style": "dark",
          "tags": Array [],
          "templating": Object {
            "list": Array [],
          },
          "time": Object {
            "from": "now-6h",
            "to": "now",
          },
          "timepicker": Object {},
          "timezone": "",
          "title": "My dashboard",
          "uid": null,
          "version": 0,
        }
      }
      isInView={false}
      panel={
        PanelModel {
          "cachedPluginOptions": Object {},
          "datasource": null,
          "events": Emitter {
            "emitter": EventEmitter {
              "_events": Object {},
              "_eventsCount": 0,
            },
          },
          "gridPos": Object {
            "h": 10,
            "w": 24,
            "x": 0,
            "y": 0,
          },
          "id": 1,
          "isInView": false,
          "targets": Array [
            Object {
              "refId": "A",
            },
          ],
          "title": "My graph",
          "transparent": false,
          "type": "graph",
        }
      }
    />
  </div>
  <div
    className=""
    id="panel-2"
    key="2"
  >
    <DashboardPanel
      dashboard={
        DashboardModel {
          "annotations": Object {
            "list": Array [
              Object {
                "builtIn": 1,
                "datasource": "-- Grafana --",
                "enable": true,
                "hide": true,
                "iconColor": "rgba(0, 211, 255, 1)",
                "name": "Annotations & Alerts",
                "type": "dashboard",
              },
            ],
          },
          "autoUpdate": undefined,
          "description": undefined,
          "editable": true,
          "events": Emitter {
            "emitter": EventEmitter {
              "_events": Object {
                "panel-added": EE {
                  "context": [Circular],
                  "fn": [Function],
                  "once": false,
                },
                "panel-removed": EE {
                  "context": [Circular],
                  "fn": [Function],
                  "once": false,
                },
                "repeats-processed": EE {
                  "context": [Circular],
                  "fn": [Function],
                  "once": false,
                },
                "row-collapsed": EE {
                  "context": [Circular],
                  "fn": [Function],
                  "once": false,
                },
                "row-expanded": EE {
                  "context": [Circular],
                  "fn": [Function],
                  "once": false,
                },
                "view-mode-changed": EE {
                  "context": [Circular],
                  "fn": [Function],
                  "once": false,
                },
              },
              "_eventsCount": 6,
            },
          },
          "gnetId": null,
          "graphTooltip": 0,
          "id": null,
          "links": Array [],
          "meta": Object {
            "canEdit": true,
            "canMakeEditable": false,
            "canSave": true,
            "canShare": true,
            "canStar": true,
            "fullscreen": false,
            "isEditing": false,
            "showSettings": true,
          },
          "originalTemplating": Array [],
          "originalTime": Object {
            "from": "now-6h",
            "to": "now",
          },
          "panels": Array [
            PanelModel {
              "cachedPluginOptions": Object {},
              "datasource": null,
              "events": Emitter {
                "emitter": EventEmitter {
                  "_events": Object {},
                  "_eventsCount": 0,
                },
              },
              "gridPos": Object {
                "h": 10,
                "w": 24,
                "x": 0,
                "y": 0,
              },
              "id": 1,
              "isInView": false,
              "targets": Array [
                Object {
                  "refId": "A",
                },
              ],
              "title": "My graph",
              "transparent": false,
              "type": "graph",
            },
            PanelModel {
              "cachedPluginOptions": Object {},
              "datasource": null,
              "events": Emitter {
                "emitter": EventEmitter {
                  "_events": Object {},
                  "_eventsCount": 0,
                },
              },
              "gridPos": Object {
                "h": 10,
                "w": 25,
                "x": 0,
                "y": 10,
              },
              "id": 2,
              "isInView": false,
              "targets": Array [
                Object {
                  "refId": "A",
                },
              ],
              "title": "My graph2",
              "transparent": false,
              "type": "graph2",
            },
            PanelModel {
              "cachedPluginOptions": Object {},
              "datasource": null,
              "events": Emitter {
                "emitter": EventEmitter {
                  "_events": Object {},
                  "_eventsCount": 0,
                },
              },
              "gridPos": Object {
                "h": 100,
                "w": 25,
                "x": 0,
                "y": 20,
              },
              "id": 3,
              "isInView": false,
              "targets": Array [
                Object {
                  "refId": "A",
                },
              ],
              "title": "My graph3",
              "transparent": false,
              "type": "graph3",
            },
            PanelModel {
              "cachedPluginOptions": Object {},
              "datasource": null,
              "events": Emitter {
                "emitter": EventEmitter {
                  "_events": Object {},
                  "_eventsCount": 0,
                },
              },
              "gridPos": Object {
                "h": 10,
                "w": 25,
                "x": 0,
                "y": 120,
              },
              "id": 4,
              "isInView": false,
              "targets": Array [
                Object {
                  "refId": "A",
                },
              ],
              "title": "My graph4",
              "transparent": false,
              "type": "graph4",
            },
          ],
          "refresh": undefined,
          "revision": undefined,
          "schemaVersion": 22,
          "snapshot": undefined,
          "style": "dark",
          "tags": Array [],
          "templating": Object {
            "list": Array [],
          },
          "time": Object {
            "from": "now-6h",
            "to": "now",
          },
          "timepicker": Object {},
          "timezone": "",
          "title": "My dashboard",
          "uid": null,
          "version": 0,
        }
      }
      isInView={false}
      panel={
        PanelModel {
          "cachedPluginOptions": Object {},
          "datasource": null,
          "events": Emitter {
            "emitter": EventEmitter {
              "_events": Object {},
              "_eventsCount": 0,
            },
          },
          "gridPos": Object {
            "h": 10,
            "w": 25,
            "x": 0,
            "y": 10,
          },
          "id": 2,
          "isInView": false,
          "targets": Array [
            Object {
              "refId": "A",
            },
          ],
          "title": "My graph2",
          "transparent": false,
          "type": "graph2",
        }
      }
    />
  </div>
  <div
    className=""
    id="panel-3"
    key="3"
  >
    <DashboardPanel
      dashboard={
        DashboardModel {
          "annotations": Object {
            "list": Array [
              Object {
                "builtIn": 1,
                "datasource": "-- Grafana --",
                "enable": true,
                "hide": true,
                "iconColor": "rgba(0, 211, 255, 1)",
                "name": "Annotations & Alerts",
                "type": "dashboard",
              },
            ],
          },
          "autoUpdate": undefined,
          "description": undefined,
          "editable": true,
          "events": Emitter {
            "emitter": EventEmitter {
              "_events": Object {
                "panel-added": EE {
                  "context": [Circular],
                  "fn": [Function],
                  "once": false,
                },
                "panel-removed": EE {
                  "context": [Circular],
                  "fn": [Function],
                  "once": false,
                },
                "repeats-processed": EE {
                  "context": [Circular],
                  "fn": [Function],
                  "once": false,
                },
                "row-collapsed": EE {
                  "context": [Circular],
                  "fn": [Function],
                  "once": false,
                },
                "row-expanded": EE {
                  "context": [Circular],
                  "fn": [Function],
                  "once": false,
                },
                "view-mode-changed": EE {
                  "context": [Circular],
                  "fn": [Function],
                  "once": false,
                },
              },
              "_eventsCount": 6,
            },
          },
          "gnetId": null,
          "graphTooltip": 0,
          "id": null,
          "links": Array [],
          "meta": Object {
            "canEdit": true,
            "canMakeEditable": false,
            "canSave": true,
            "canShare": true,
            "canStar": true,
            "fullscreen": false,
            "isEditing": false,
            "showSettings": true,
          },
          "originalTemplating": Array [],
          "originalTime": Object {
            "from": "now-6h",
            "to": "now",
          },
          "panels": Array [
            PanelModel {
              "cachedPluginOptions": Object {},
              "datasource": null,
              "events": Emitter {
                "emitter": EventEmitter {
                  "_events": Object {},
                  "_eventsCount": 0,
                },
              },
              "gridPos": Object {
                "h": 10,
                "w": 24,
                "x": 0,
                "y": 0,
              },
              "id": 1,
              "isInView": false,
              "targets": Array [
                Object {
                  "refId": "A",
                },
              ],
              "title": "My graph",
              "transparent": false,
              "type": "graph",
            },
            PanelModel {
              "cachedPluginOptions": Object {},
              "datasource": null,
              "events": Emitter {
                "emitter": EventEmitter {
                  "_events": Object {},
                  "_eventsCount": 0,
                },
              },
              "gridPos": Object {
                "h": 10,
                "w": 25,
                "x": 0,
                "y": 10,
              },
              "id": 2,
              "isInView": false,
              "targets": Array [
                Object {
                  "refId": "A",
                },
              ],
              "title": "My graph2",
              "transparent": false,
              "type": "graph2",
            },
            PanelModel {
              "cachedPluginOptions": Object {},
              "datasource": null,
              "events": Emitter {
                "emitter": EventEmitter {
                  "_events": Object {},
                  "_eventsCount": 0,
                },
              },
              "gridPos": Object {
                "h": 100,
                "w": 25,
                "x": 0,
                "y": 20,
              },
              "id": 3,
              "isInView": false,
              "targets": Array [
                Object {
                  "refId": "A",
                },
              ],
              "title": "My graph3",
              "transparent": false,
              "type": "graph3",
            },
            PanelModel {
              "cachedPluginOptions": Object {},
              "datasource": null,
              "events": Emitter {
                "emitter": EventEmitter {
                  "_events": Object {},
                  "_eventsCount": 0,
                },
              },
              "gridPos": Object {
                "h": 10,
                "w": 25,
                "x": 0,
                "y": 120,
              },
              "id": 4,
              "isInView": false,
              "targets": Array [
                Object {
                  "refId": "A",
                },
              ],
              "title": "My graph4",
              "transparent": false,
              "type": "graph4",
            },
          ],
          "refresh": undefined,
          "revision": undefined,
          "schemaVersion": 22,
          "snapshot": undefined,
          "style": "dark",
          "tags": Array [],
          "templating": Object {
            "list": Array [],
          },
          "time": Object {
            "from": "now-6h",
            "to": "now",
          },
          "timepicker": Object {},
          "timezone": "",
          "title": "My dashboard",
          "uid": null,
          "version": 0,
        }
      }
      isInView={false}
      panel={
        PanelModel {
          "cachedPluginOptions": Object {},
          "datasource": null,
          "events": Emitter {
            "emitter": EventEmitter {
              "_events": Object {},
              "_eventsCount": 0,
            },
          },
          "gridPos": Object {
            "h": 100,
            "w": 25,
            "x": 0,
            "y": 20,
          },
          "id": 3,
          "isInView": false,
          "targets": Array [
            Object {
              "refId": "A",
            },
          ],
          "title": "My graph3",
          "transparent": false,
          "type": "graph3",
        }
      }
    />
  </div>
  <div
    className=""
    id="panel-4"
    key="4"
  >
    <DashboardPanel
      dashboard={
        DashboardModel {
          "annotations": Object {
            "list": Array [
              Object {
                "builtIn": 1,
                "datasource": "-- Grafana --",
                "enable": true,
                "hide": true,
                "iconColor": "rgba(0, 211, 255, 1)",
                "name": "Annotations & Alerts",
                "type": "dashboard",
              },
            ],
          },
          "autoUpdate": undefined,
          "description": undefined,
          "editable": true,
          "events": Emitter {
            "emitter": EventEmitter {
              "_events": Object {
                "panel-added": EE {
                  "context": [Circular],
                  "fn": [Function],
                  "once": false,
                },
                "panel-removed": EE {
                  "context": [Circular],
                  "fn": [Function],
                  "once": false,
                },
                "repeats-processed": EE {
                  "context": [Circular],
                  "fn": [Function],
                  "once": false,
                },
                "row-collapsed": EE {
                  "context": [Circular],
                  "fn": [Function],
                  "once": false,
                },
                "row-expanded": EE {
                  "context": [Circular],
                  "fn": [Function],
                  "once": false,
                },
                "view-mode-changed": EE {
                  "context": [Circular],
                  "fn": [Function],
                  "once": false,
                },
              },
              "_eventsCount": 6,
            },
          },
          "gnetId": null,
          "graphTooltip": 0,
          "id": null,
          "links": Array [],
          "meta": Object {
            "canEdit": true,
            "canMakeEditable": false,
            "canSave": true,
            "canShare": true,
            "canStar": true,
            "fullscreen": false,
            "isEditing": false,
            "showSettings": true,
          },
          "originalTemplating": Array [],
          "originalTime": Object {
            "from": "now-6h",
            "to": "now",
          },
          "panels": Array [
            PanelModel {
              "cachedPluginOptions": Object {},
              "datasource": null,
              "events": Emitter {
                "emitter": EventEmitter {
                  "_events": Object {},
                  "_eventsCount": 0,
                },
              },
              "gridPos": Object {
                "h": 10,
                "w": 24,
                "x": 0,
                "y": 0,
              },
              "id": 1,
              "isInView": false,
              "targets": Array [
                Object {
                  "refId": "A",
                },
              ],
              "title": "My graph",
              "transparent": false,
              "type": "graph",
            },
            PanelModel {
              "cachedPluginOptions": Object {},
              "datasource": null,
              "events": Emitter {
                "emitter": EventEmitter {
                  "_events": Object {},
                  "_eventsCount": 0,
                },
              },
              "gridPos": Object {
                "h": 10,
                "w": 25,
                "x": 0,
                "y": 10,
              },
              "id": 2,
              "isInView": false,
              "targets": Array [
                Object {
                  "refId": "A",
                },
              ],
              "title": "My graph2",
              "transparent": false,
              "type": "graph2",
            },
            PanelModel {
              "cachedPluginOptions": Object {},
              "datasource": null,
              "events": Emitter {
                "emitter": EventEmitter {
                  "_events": Object {},
                  "_eventsCount": 0,
                },
              },
              "gridPos": Object {
                "h": 100,
                "w": 25,
                "x": 0,
                "y": 20,
              },
              "id": 3,
              "isInView": false,
              "targets": Array [
                Object {
                  "refId": "A",
                },
              ],
              "title": "My graph3",
              "transparent": false,
              "type": "graph3",
            },
            PanelModel {
              "cachedPluginOptions": Object {},
              "datasource": null,
              "events": Emitter {
                "emitter": EventEmitter {
                  "_events": Object {},
                  "_eventsCount": 0,
                },
              },
              "gridPos": Object {
                "h": 10,
                "w": 25,
                "x": 0,
                "y": 120,
              },
              "id": 4,
              "isInView": false,
              "targets": Array [
                Object {
                  "refId": "A",
                },
              ],
              "title": "My graph4",
              "transparent": false,
              "type": "graph4",
            },
          ],
          "refresh": undefined,
          "revision": undefined,
          "schemaVersion": 22,
          "snapshot": undefined,
          "style": "dark",
          "tags": Array [],
          "templating": Object {
            "list": Array [],
          },
          "time": Object {
            "from": "now-6h",
            "to": "now",
          },
          "timepicker": Object {},
          "timezone": "",
          "title": "My dashboard",
          "uid": null,
          "version": 0,
        }
      }
      isInView={false}
      panel={
        PanelModel {
          "cachedPluginOptions": Object {},
          "datasource": null,
          "events": Emitter {
            "emitter": EventEmitter {
              "_events": Object {},
              "_eventsCount": 0,
            },
          },
          "gridPos": Object {
            "h": 10,
            "w": 25,
            "x": 0,
            "y": 120,
          },
          "id": 4,
          "isInView": false,
          "targets": Array [
            Object {
              "refId": "A",
            },
          ],
          "title": "My graph4",
          "transparent": false,
          "type": "graph4",
        }
      }
    />
  </div>
</SizeMe(GridWrapper)>
`;
